### ISSUE TITLE
Host memory leak in `to_rarray` with a sharding annotation (multi-device mesh)

### ISSUE THREAD

**Issue by csvance** [EnzymeAD/Reactant.jl#2948]
## Summary

Calling `Reactant.to_rarray(x; sharding=...)` in a loop on a multi-device mesh
leaks host memory: roughly a constant ~2 MB per call, independent of the array
size and of whether the `Sharding` object is reused across calls. The Julia heap
(`Base.gc_live_bytes()`) stays flat and a `GC.gc()` every iteration does not
reclaim the growth, so the leaked memory is external to Julia (XLA/PJRT host
allocations backing the per-device buffers/transfers).

The identical loop without a sharding annotation does not leak. This reproduces
on the **CPU** backend with a 4-device host mesh, so it is not GPU- or
driver-specific.

In a training loop that materializes sharded inputs and optimizer scalars every
step (~15 sharded `to_rarray` calls/step), RSS climbed unbounded to ~64 GB.

## Reproducer

Self-contained, no GPU required. Run with a 4-device host mesh:

```
XLA_FLAGS=--xla_force_host_platform_device_count=4 julia --project=. mwe.jl
```

All setup and the hot loop are inside `main()` (no non-const globals are read in
the loop), so this is not a global-scope / type-instability artifact.

```julia
using Reactant
using Reactant: Sharding

rss_gb()  = parse(Int, split(read("/proc/self/statm", String))[2]) * 4096 / 2^30
live_gb() = Base.gc_live_bytes() / 2^30

function run_loop(label, make_rarray)
    println("== $label =="); flush(stdout)
    for it in 1:40
        for _ in 1:20
            make_rarray()
        end
        GC.gc()
        it % 10 == 0 && println("  iter=$it rss_gb=", round(rss_gb(); digits=3),
            " julia_live_gb=", round(live_gb(); digits=3))
        flush(stdout)
    end
end

function main()
    devs = Reactant.devices()
    @assert length(devs) >= 2 "needs >= 2 devices"
    mesh = Sharding.Mesh(reshape(collect(devs), length(devs)), (:data,))
    sharding = Sharding.NamedSharding(mesh, (nothing, :data))
    println("devices=", length(devs))
    run_loop("sharded to_rarray (leaks)",
        () -> Reactant.to_rarray(randn(Float32, 256, 64); sharding = sharding))
    run_loop("unsharded to_rarray (control)",
        () -> Reactant.to_rarray(randn(Float32, 256, 64)))
end

main()
```

## Observed output (CPU, 4 devices)

```
devices=4
== sharded to_rarray (leaks) ==
  iter=10 rss_gb=1.885 julia_live_gb=0.04
  iter=20 rss_gb=2.314 julia_live_gb=0.04
  iter=30 rss_gb=2.743 julia_live_gb=0.04
  iter=40 rss_gb=3.172 julia_live_gb=0.04
== unsharded to_rarray (control) ==
  iter=10 rss_gb=3.174 julia_live_gb=0.04
  iter=20 rss_gb=3.174 julia_live_gb=0.04
  iter=30 rss_gb=3.174 julia_live_gb=0.04
  iter=40 rss_gb=3.174 julia_live_gb=0.04
```

The sharded loop grows ~43 MB/iter (20 calls of a 64 KB array, so ~2.1 MB leaked
per call, independent of the 64 KB payload). The unsharded control, run in the
same process right after, is pinned at 3.174 GB across all 40 iterations. The
Julia heap is flat (0.04 GB) throughout.

## Notes / what was ruled out

- Not the Julia heap: `gc_live_bytes()` is flat and `GC.gc()` runs every
  iteration without reclaiming the growth.
- Not array-size dependent: the per-call leak is a fixed overhead (~2 MB), much
  larger than the 64 KB payload.
- Not `Sharding`-object churn: caching/reusing a single `NamedSharding` across
  all calls (as above) leaks the same amount as constructing one per call.
- Sharding-specific: the unsharded control loop in the same process does not
  leak at all.
- Not a global-scope / type-instability artifact: the reproducer runs entirely
  inside `main()` with no non-const globals read in the loop, and the leak is
  external memory (the Julia heap is flat), not GC-reclaimable Julia boxing.
- Not reclaimable by eager buffer finalization: running `Base.finalize` on each
  per-device `AsyncBuffer.buffer` (and its `future`) of the returned array,
  i.e. forcing the registered `XLA.free_buffer` finalizer immediately, leaks at
  the identical rate (RSS slope unchanged). The leaked memory is therefore not
  in the per-device buffers (those finalize correctly) but in auxiliary host
  state allocated during sharded construction that has no Julia-side handle or
  finalizer. There is no user-side mitigation.

## Environment

- Reactant 0.2.262, Reactant_jll 0.0.383
- Enzyme 0.13.149, Lux 1.31.4
- Julia 1.12.6
- Reproduces on the CPU backend (`--xla_force_host_platform_device_count=4`).
  Originally observed on 4 × NVIDIA RTX A6000 (driver 580.159.03) during
  multi-GPU training.

**Comment by csvance** [comment]
## Root cause

The eager sharded path builds an `HloSharding` via `Base.convert(::Type{HloSharding}, ::NamedSharding)` in `src/Sharding.jl`. When no MLIR context was active, that function created a fresh `Reactant.ReactantContext()`, an MLIR `Context` with every dialect registered (a multi-MB C++ object), and ran the conversion inside `MLIR.IR.@with_context`.

The problem: `@with_context` only `activate`s and `deactivate`s the context (a task-local stack push/pop). It never disposes it. MLIR `Context` uses manual `mark_alloc`/`dispose` ownership with **no GC finalizer** (`src/mlir/IR/Context.jl`), so a context that is created, activated, then merely deactivated and dropped is leaked. One leaked context per call accounts for the fixed ~2 MB, independent of array size (the cost is the dialect-registration, not the data).

## Fix

Dispose the context when, and only when, we created it. `convert` now captures `has_ctx = has_context()` and disposes the context in a `finally` only when `has_ctx` is false; the padding branch disposes unconditionally (its context is always freshly created). This is safe because the returned `XLA.HloSharding` is a standalone C++ object that holds no reference back into the MLIR context.
